### PR TITLE
[WIP] Fix issue 22502

### PR DESCRIPTION
**Pull request by cshung**

### CODE REVIEW

File: src/vm/eventpipebuffermanager.h [hunk 1]
```diff
@@ -141,8 +141,8 @@ class EventPipeBufferList
     // If it is false, then this buffer can be de-allocated after it is drained.
     Volatile<bool> m_ownedByThread;
 
-    // True if a thread is writing something to this list
-    Volatile<bool> m_threadEventWriteInProgress;
+    // Lock to protect access to the active buffer
```
Comment (sywhang) [inline src/vm/eventpipebuffermanager.h]: We might want to be more specific about what "active buffer" is referring to for the unfortunate soul that needs to touch this part of the runtime after us.

**Comment by sywhang** [comment]
Just a heads up but I think we should do a better job documenting the code. When I first started digging into this code I had to spend a lot of time trying to understand what lock is trying to protect what because it isn't immediately obvious (and I'm assuming you did too). Now that we're putting another lock I think we should put more comments that makes it obvious what we are trying to do with these locks.

File: src/vm/eventpipebuffermanager.cpp [hunk 2]
```diff
@@ -353,14 +353,12 @@ bool EventPipeBufferManager::WriteEvent(Thread *pThread, EventPipeSession &sessi
...
-
-            // Attempt to write the event to the buffer.  If this fails, we should allocate a new buffer.
-            allocNewBuffer = !pBuffer->WriteEvent(pEventThread, session, event, payload, pActivityId, pRelatedActivityId, pStack);
+            {
+                SpinLockHolder _slh(pThreadBufferList->GetLock());
 
-            // Mark that the thread is no longer writing an event.
-            pThreadBufferList->SetThreadEventWriteInProgress(false);
+                // Attempt to write the event to the buffer.  If this fails, we should allocate a new buffer.
+                allocNewBuffer = !pBuffer->WriteEvent(pEventThread, session, event, payload, pActivityId, pRelatedActivityId, pStack);
```
Comment (cshung) [inline src/vm/eventpipebuffermanager.cpp]: I just realized that this call could lead to doing a full stackwalk - we need to figure out a way to make the spin lock region much smaller than that.

Comment (cshung) [reply src/vm/eventpipebuffermanager.cpp]: This is still work in progress, with #23185, we were able to move stack walk out of the critical section. 

We still have a couple of issues remaining:

- [ ] SpinLock nesting - right now I just commented out the assert to allow SpinLock nest, which is not good.
- [ ] SpinLock being disposed of while we are still holding the lock in the DeAllocate() code path

The plan to solve the SpinLock nesting problem is to turn EventPipeBufferManager::m_lock into a critical section, and the plan to solve the SpinLock being disposed problem is to store the SpinLock itself somewhere else and never dispose it.

**Comment by cshung** [comment]
This is obsolete, closed.

**Pull request closed without merging**